# <p style="text-align: center">Финальный проект. Работа с SQL.<p>

### Опиcание цели исследования

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.



## <p style="text-align: center"> <a id="soder">СОДЕРЖАНИЕ</a> </p>

[Импорт библиотек. Подключение к БД. Просмотр таблиц. ](#task0) 

[Задание 1](#task1)  Посчитайте, сколько книг вышло после 1 января 2000 года;


[Задание 2](#task2)  Для каждой книги посчитайте количество обзоров и среднюю оценку;


[Задание 3](#task3)  Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;


[Задание 4](#task4)  Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;


[Задание 5](#task5)  Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


## <a id="task0"> Импорт библиотек. Подключение к БД. Просмотр таблиц.  </a>  
[К содержанию](#soder)

* Исследуйте таблицы — выведите первые строки;

In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-

# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
import datetime as dt



In [2]:
# устанавливаем параметры 
db_config = {
    'user': 'praktikum_student',  # имя пользователя 
    'pwd': 'Sdf4$2;d-d30pp',      # пароль 
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432,                 # порт подключения 
    'db': 'data-analyst-final-project-db'  # название базы данных 
} 

connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config) 

# сохраняем коннектор 
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

# создаем список таблиц для запроса
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

# выполняем SQL-запрос для каждой таблицы и сохраняем результаты в DataFrames
dataframes = {}
for table in tables:
    query = f'SELECT * FROM {table}'
    con = engine.connect()
    dataframes[table] = pd.io.sql.read_sql(sql=text(query), con=con)

# закрываем соединение
con.close()

# выводим данные из всех таблиц
for table, df in dataframes.items():
    print(f"\nТаблица - {table}")
    print(df)


Таблица - books
     book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id 

С первого взгляда, данные в порядке, пропусков нет. (Можно поменять тип в колонке publication_date - на дату)

**Для проверки запишем каждую таблицу в ДатаФрейм**

In [3]:
query_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(query_books, con = engine)
# books

query_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
# reviews

query_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
# ratings

query_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
# publishers

query_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(query_authors, con = engine)
# authors

## Задание 1 <a id="task1"></a>  
[К содержанию](#soder)  

* Посчитайте, сколько книг вышло после 1 января 2000 года;

In [4]:
query1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

request1 = pd.io.sql.read_sql(query1, con = engine)
# request1

print(f'После 1 января 2000 года вышло {request1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


***ВЫВОД: После 1 января 2000 года вышло 819 книг***

*ПРОВЕРКА*

In [5]:
# Преобразуем столбец 'publication_date' в формат datetime
books['publication_date'] = pd.to_datetime(books['publication_date'])

# Считаем количество книг, опубликованных после 2000-01-01
count_books_after_2000 = books[books['publication_date'] > '2000-01-01']['book_id'].count()

print(count_books_after_2000)

819


## Задание 2 <a id="task2"></a>  
[К содержанию](#soder)  

* Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [6]:
query2 = ''' 

SELECT 
    books.title AS book_name,
    books.book_id AS id,
    SubRev.n_rev AS reviews,
    SubRat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS SubRat ON SubRat.book_id = books.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS SubRev ON SubRev.book_id = books.book_id
                
; ''' 

request2 = pd.io.sql.read_sql(query2, con = engine)
request2

,book_name,id,reviews,avg_rating
0,The Body in the Library (Miss Marple #3),652,2.0,4.500000
1,Galápagos,273,2.0,4.500000
2,A Tree Grows in Brooklyn,51,5.0,4.250000
3,Undaunted Courage: The Pioneering First Missio...,951,2.0,4.000000
4,The Prophet,839,4.0,4.285714
...,...,...,...,...
995,Alice in Wonderland,64,4.0,4.230769
996,A Woman of Substance (Emma Harte Saga #1),55,2.0,5.000000
997,Christine,148,3.0,3.428571
998,The Magicians' Guild (Black Magician Trilogy #1),790,2.0,3.500000


### Результат вывода SQL-запроса представляет собой таблицу, в которой для каждой книги из базы данных указаны следующие параметры:

**Название книги (book_name):**

- Название книги, как указано в таблице books.
- Пример: "The Body in the Library (Miss Marple #3)".

**Идентификатор книги (id):**

- Уникальный числовой идентификатор книги из таблицы books.
- Пример: 652.

**Количество обзоров (reviews):**

- Число обзоров, оставленных пользователями для конкретной книги (данные из таблицы reviews).
- Если книга не имеет обзоров, значение равно NULL или 0.
- Пример: *2 обзора для книги "The Body in the Library (Miss Marple #3)".

**Средняя оценка (avg_rating):**

- Средняя оценка книги, рассчитанная на основе данных из таблицы ratings.
- Если у книги нет оценок, значение равно NULL.
- Пример: Средняя оценка для книги "The Body in the Library (Miss Marple #3)" составляет 4.5.

---  
*ПРОВЕРКА*

In [7]:
r2 = books[['title','book_id']].merge(reviews.groupby('book_id')['review_id'].count().reset_index(), 
                                 how='left', on='book_id').merge(ratings.groupby('book_id')['rating'].mean().reset_index(), how='left', on='book_id')
r2.query('book_id == [652,273,51,951,839,64,55,148,790,828]')
# r2[r2['book_id'].isin([652,273,51,951,839])]

,title,book_id,review_id,rating
50,A Tree Grows in Brooklyn,51,5.0,4.250000
54,A Woman of Substance (Emma Harte Saga #1),55,2.0,5.000000
63,Alice in Wonderland,64,4.0,4.230769
147,Christine,148,3.0,3.428571
272,Galápagos,273,2.0,4.500000
651,The Body in the Library (Miss Marple #3),652,2.0,4.500000
789,The Magicians' Guild (Black Magician Trilogy #1),790,2.0,3.500000
827,The Plot Against America,828,2.0,3.000000
838,The Prophet,839,4.0,4.285714
950,Undaunted Courage: The Pioneering First Missio...,951,2.0,4.000000


## Задание 3 <a id="task3"></a>  
[К содержанию](#soder)  

* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [8]:
query3 = ''' 
SELECT
    T.publisher_id,
    T.publisher,
    T.n_books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS n_books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        n_books DESC) AS T
LIMIT 1
; ''' 

request3 = pd.io.sql.read_sql(query3, con = engine)
request3

,publisher_id,publisher,n_books
0,212,Penguin Books,42


***ВЫВОД: Издательство, которое выпустило наибольшее число книг, 42 книги, толще 50 страниц - это "Penguin Books"***  

*ПРОВЕРКА*

In [9]:
books.query('num_pages>50')[['publisher_id','num_pages', 'book_id']]. \
        groupby('publisher_id')['book_id'].count().reset_index().sort_values(by='book_id',ascending=False)

,publisher_id,book_id
207,212,42
302,309,31
111,116,25
212,217,24
31,33,19
...,...,...
83,86,1
199,204,1
200,205,1
82,85,1


In [10]:
publishers.query('publisher_id==212')

,publisher_id,publisher
211,212,Penguin Books


## Задание 4 <a id="task4"></a>  
[К содержанию](#soder)  

* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [11]:
query4 = ''' 
SELECT 
    aut_avg.author,
    AVG(aut_avg.avg_rating) AS avg_avg_rating
FROM
        (SELECT 
            authors.author AS author
           , AVG(ratings.rating) AS avg_rating
        FROM books
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY 
             books.book_id
             , authors.author_id
        HAVING
            COUNT(ratings.rating_id) >=50 
        ) AS aut_avg 
GROUP BY    
    aut_avg.author
ORDER BY
    avg_avg_rating DESC
LIMIT 5
    
; ''' 

request4 = pd.io.sql.read_sql(query4, con = engine)
request4

,author,avg_avg_rating
0,J.K. Rowling/Mary GrandPré,4.283844
1,Markus Zusak/Cao Xuân Việt Khương,4.264151
2,J.R.R. Tolkien,4.258446
3,Louisa May Alcott,4.192308
4,Rick Riordan,4.080645


***Самую высокую оценку получает - J.K. Rowling/Mary GrandPré - 4.283844 (Гарри Поттер ФОРЕВЕР)***

### Вариант 2

In [12]:
query4 = ''' 
SELECT 
    authors.author AS author,
    AVG(sub.avg_rating) AS avg_author_rating
FROM (
        SELECT 
            books.author_id AS author_id,
            AVG(ratings.rating) AS avg_rating
        FROM books
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY books.book_id
        HAVING COUNT(ratings.rating_id) >= 50
    ) AS sub
LEFT JOIN authors ON sub.author_id = authors.author_id
GROUP BY authors.author
ORDER BY avg_author_rating DESC
LIMIT 5;
''' 

request4 = pd.io.sql.read_sql(query4, con=engine)
print(request4)



                              author  avg_author_rating
0         J.K. Rowling/Mary GrandPré           4.283844
1  Markus Zusak/Cao Xuân Việt Khương           4.264151
2                     J.R.R. Tolkien           4.258446
3                  Louisa May Alcott           4.192308
4                       Rick Riordan           4.080645


***Ответ тот же, Самую высокую оценку получает - J.K. Rowling/Mary GrandPré - 4.283844 (Гарри Поттер ФОРЕВЕР)***

## Задание 5 <a id="task5"></a>  
[К содержанию](#soder)  
 * Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

***Из предоставленных данных непонятно (неочевидно) являются ли пользователи из таблиц `ratings` и `reviews` одними и теми же, поэтому надо это проверить.***

In [13]:
print('Кол-во пользователей в "reviews" =', reviews['username'].nunique())
print('Кол-во пользователей в "ratings" =', ratings['username'].nunique())

Кол-во пользователей в "reviews" = 160
Кол-во пользователей в "ratings" = 160


**По кол-ву совпадает, но проверим еще по множествам**

In [14]:
set(reviews['username'].unique().tolist()) == set(ratings['username'].unique().tolist())

True

***Множества идентичны. Всё в порядке, это одни и теже пользователи. Можно составлять запрос.***

In [15]:
query5 = ''' 

SELECT
    AVG(Sub.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 48
                    )
    GROUP BY 
        username
    ) AS Sub
; ''' 

request5 = pd.io.sql.read_sql(query5, con = engine)
request5

,avg
0,24.0


***Итак среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24***

*Проверим*

In [16]:
users = ratings.groupby('username')['rating_id'].count().reset_index()
users = users.query('rating_id > 48')
users

,username,rating_id
6,andreaeaton,49
57,jennifermiller,53
76,lesliegibbs,50
83,martinadam,56
104,paul88,56
110,richard89,55
123,sfitzgerald,55
127,shermannatalie,50
130,susan85,49
144,vanessagardner,50


In [17]:
users_lst = users["username"].tolist()
# reviews.query('username.isin(@users_lst)')
reviews.query('username in @users_lst').groupby('username')['review_id'].count().mean()

24.0

**Общий вывод:**

- Книги, вышедшие после 1 января 2000 года:
- После указанной даты было опубликовано 819 книг, что демонстрирует значительный вклад современных издательств в литературу.

- Издательство с наибольшим числом книг толще 50 страниц:
Лидером стало издательство "Penguin Books", выпустившее 42 книги такого формата. Это указывает на их ориентацию на публикацию объемных и, вероятно, качественных произведений.

- Автор с самой высокой средней оценкой:
Самую высокую среднюю оценку книг — 4.283844 — получил дуэт авторов J.K. Rowling/Mary GrandPré, благодаря их легендарной серии про Гарри Поттера. Безусловно, магия этой серии продолжает вдохновлять миллионы читателей по всему миру.

- Среднее количество обзоров от активных пользователей:
Пользователи, оставившие более 48 оценок, в среднем написали 24 обзора, что демонстрирует высокий уровень вовлеченности активной аудитории.

Эти результаты дают ценное представление о литературных предпочтениях, влиянии современных авторов и издательств, а также о вовлеченности читателей. Гарри Поттер навсегда останется культовым феноменом! С наступающим новым годом!